### PR TITLE
refactor(module: graph): refactor some apis and support to disable animations

### PR DESCRIPTION
**Pull request by simplejason**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

**Break Changes**
| Before Method | Current |
| --- | --- |
| autoFit() | fitCenter() |
| nzShowArrow | removed(Default is shown) |

**Features**
- support custom edges
- support `nzNoAnimation`

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] Application (the showcase website) / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?
```
[x] Yes
[ ] No
```

support zooming and disabling animations
```html
<nz-graph nzNoAnimation nz-graph-zoom [nzGraphData]="data">
</nz-graph>
```

> Problems still not resolved

- performance of animations
- minimap support(including performance and styles)

### CODE REVIEW

**Comment by simplejason** [comment]
@Airblader could you please show us how you use the graph component? If the problem still exists, we can fix it in this PR.

**Review by Airblader — COMMENTED** [review]
@simplejason Thanks for your work on this! I've left a couple of comments, also some more points: 

* Being able to disable the animations will help us, but is ultimately also just a workaround. What would be nice if there was a way to update nodes without updating the graph topology (i.e. no new/removed nodes or edges) while still having animations when the topology does change.
* It would also be nice to have an API to re-render / re-layout the graph. I need this for nodes which are interactive and can change in size when the user interacts with it. I currently trigger the re-render by calling setData again. Even cooler would be if nz-graph observed size changes of nodes itself and updated the graph accordingly, but I'm not sure how easily that can be achieved, so an explicit re-render / re-layout API would also be OK.
* The auto-fit behavior currently doesn't always work perfectly for me. In general it provides good results, but sometimes one node is slightly cut off at the bottom.
* In general it would of course be really awesome if the user was able to provide their own graph layout algorithm, but we don't need this at the moment.

Animations in the nodes seem to be a real problem for the graph at the moment. If you have a button inside a node, clicking it will mysteriously cause other nodes to partially or fully disappear while the button's "wave" animation is playing. If you use a button without animation (i.e. a native HTML button without nz-button, or an nz-button with nzType link) this issue does not occur. Using nzNoAnimation on the button also fixes it.

Another issue I ran into was wanting to disable the scroll-to-zoom behavior on a specific part of the node (because the node itself is scrollable there). I did eventually figure out that this is possible by registering an event listener in capture mode, though, which I think is an OK to solution since this is probably an edge-case problem. But initially this made me want to have more access to the underlying d3 zoom controller etc. I'm not sure how much you want to expose these things on the component API, though.

File: components/graph/graph-zoom.directive.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+  unbind(): void {
+    // Destroy listener
+    this.svgSelection?.interrupt().selectAll('*').interrupt();
+    if (this.zoomBehavior) {
+      this.zoomBehavior.on('end', null).on('zoom', null);
+    }
+  }
+
+  // Methods
+  fitCenter(duration: number = 0): void {
```
[the remark this answers is not in the document]
Comment (simplejason) [reply components/graph/graph-zoom.directive.ts:77]: `fitCenter` will move the graph to center and re-calculate again based on parent element, do we have a situation where we need to specify the scale parameter manually?

File: components/graph/graph-zoom.directive.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    this.reScale(duration);
+  }
+
+  focus(node: SVGGElement, duration: number = 0): void {
```
[the remark this answers is not in the document]
Comment (simplejason) [reply components/graph/graph-zoom.directive.ts:81]: Makes sense.

**Comment by Airblader** [comment]
> It would also be nice to have an API to re-render / re-layout the graph. I need this for nodes which are interactive and can change in size when the user interacts with it. I currently trigger the re-render by calling setData again. Even cooler would be if nz-graph observed size changes of nodes itself and updated the graph accordingly, but I'm not sure how easily that can be achieved, so an explicit re-render / re-layout API would also be OK.

FWIW, I've written this directive which will do this automatically, but you can see that it would be nice if I could just trigger a re-layout rather than calling setData, and also if nz-graph could (maybe just optionally) do this, it'd be even better.

```
/** Observes the host for resizes and relayouts the parent nz-graph when it happens. */
@Directive({
  selector: '[nodeUpdateOnResize]',
})
export class NodeUpdateOnResizeDirective implements OnInit, OnDestroy {
  private readonly destroy$ = new Subject<void>();

  constructor(
    @Optional() private readonly nzGraph: NzGraphComponent,
    private readonly elementRef: ElementRef,
    private readonly cdRef: ChangeDetectorRef
  ) {
    if (!this.nzGraph) {
      console.warn(`Failed to inject host nz-graph component.`);
    }
  }

  public ngOnInit() {
    this.createResizeObserver()
      .pipe(
        distinctUntilChanged((prev, curr) => {
          const { width: prevWidth, height: prevHeight } = prev;
          const { width: currWidth, height: currHeight } = curr;
          return prevWidth === currWidth && prevHeight === currHeight;
        }),
        debounceTime(10),
        takeUntil(this.destroy$)
      )
      .subscribe(() => {
        // FIXME: It would be nice if ng-zorro had a "trigger a re-layout" API
        this.nzGraph?.nzGraphData?.setData(this.nzGraph?.nzGraphData?.dataSource);
        this.cdRef.markForCheck();
      });
  }

  public ngOnDestroy(): void {
    this.destroy$.next();
    this.destroy$.complete();
  }

  private createResizeObserver() {
    const stream$ = new Subject<DOMRectReadOnly>();
    // Remove `as any` after https://github.com/microsoft/TypeScript/issues/37861 is merged
    // tslint:disable-next-line:no-any
    const ResizeObserver = (window as any).ResizeObserver;
    if (ResizeObserver) {
      const resizeObserver = new ResizeObserver(entries => {
        for (const entry of entries) {
          if (entry.target === this.elementRef.nativeElement) {
            stream$.next(entry.contentRect);
          }
        }
      });

      resizeObserver.observe(this.elementRef.nativeElement);
      this.destroy$.subscribe(() => resizeObserver.disconnect());
    }

    return stream$;
  }

}
```

**Comment by simplejason** [comment]
> FWIW, I've written this directive which will do this automatically, but you can see that it would be nice if I could just trigger a re-layout rather than calling setData, and also if nz-graph could (maybe just optionally) do this, it'd be even better.

@Airblader Cool, I will check your comments and I think some of thems really should be supported by the component.

File: components/graph/graph-zoom.directive.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Airblader) [reply components/graph/graph-zoom.directive.ts:77]: I guess what I want to do is have the graph fit but specify some "padding" or something to keep on the outside. Maybe the scale parameter isn't what I'm looking for, though.

**Comment by Airblader** [comment]
One more thing, when I use something like nz-tabset or nz-list inside a custom node, I need to set its z-index to 1 or something for it to show up. I'm not entirely sure why this happens (didn't check since the solution is so easy), so also not sure if anything can be done for that.

**Comment by Airblader** [comment]
@simplejason Another couple things:

nz-graph will throw a bunch of errors if the window is so small that the host element has zero dimensions. I think it should either throw a single, reasonable error, deal with the error, or, as a simpler fix, just include this:

```
nz-graph {
  // This prevents errors from being thrown if the window is so small that there's no space to render the graph
  min-height: 1px;
}
```

I've also had to add this style in order for a box-shadow on my custom nodes to work. I didn't observe any negative side effects, so not sure hidden overflow is needed in nz-graph to begin with?

```
.nz-graph-node-rect {
  // This is needed for box-shadows on the nodes to work
  overflow: visible;
}
```

And finally, when I use custom nodes, there's still styling on the nodes. I had to add the following to get rid of them so that only my custom node is visible. I'm not sure whether this is intended or not?

```
.nz-graph-node-wrapper {
  border: none;
  background: none;
}
```

**Comment by Airblader** [comment]
I'm also encountering an issue where if a node is partially off-screen and I e.g. hover over it, the browser seems to not re-render the entire object (presumably because it's not visible), and subsequently panning the node into view it hasn't rendered correctly. This is fixed once I hover over the node again after it's on-screen. I'm assuming this is more of an issue with d3 or even the browser and not ng-zorro, though. :-/

Here's two pictures to demonstrate it. One picture is what the node looks like fully rendered, the other one is what it looks like when it was rendered while partially off-screen and then panning it into view.

screenshot-2020-12-08_09-52-43
screenshot-2020-12-08_09-52-36

**Comment by Airblader** [comment]
About that directive I've shown above, I've had to modify this to work in a different way. When using this on a custom node that can expand or collapse, it only worked reliably for expand, but not collapse. The reason is that the zone becomes stable before the directive can update the data, and since nz-graph waits for the zone to be stable to auto resize nodes this won't happen (until some later update happens).

The only way I could make that work is calling detectChanges from a parent of nz-graph at least one tick after having updated the data. 

Since the design above was also suboptimal because all nodes could cause frequent concurrent updates, my directive now emits on an output instead so that these events are routed through my parent component which then performs the data update and the detectChanges call.

**Pull request merged**